src/target/riscv/riscv-013.c

```diff
@@ -5418,16 +5418,22 @@ static int riscv013_step_or_resume_current_hart(struct target *target,
 		return ERROR_OK;
 	}
 
+	LOG_TARGET_ERROR(target, "unable to %s. dmstatus=0x%08x",
+		step ? "single-step" : "resume", dmstatus);
+
 	dm_write(target, DM_DMCONTROL, dmcontrol);
+	LOG_TARGET_ERROR(target, "  cancelling RESUMEREQ");
 
-	LOG_TARGET_ERROR(target, "unable to resume");
 	if (dmstatus_read(target, &dmstatus, true) != ERROR_OK)
 		return ERROR_FAIL;
-	LOG_TARGET_ERROR(target, "  dmstatus=0x%08x", dmstatus);
+
+	LOG_TARGET_ERROR(target, "  dmstatus after cancellation=0x%08x", dmstatus);
 
 	if (step) {
-		LOG_TARGET_ERROR(target, "  was stepping, halting");
+		LOG_TARGET_ERROR(target,
+			"  trying to restore from failed single-step, by requesting throw-away halt");
```

Comment:
I have proposed minor reword for clarity.

What have you please meant by "throw-away" halt? For me the sentence makes sense without it.

```suggestion
		LOG_TARGET_ERROR(target,
			"  trying to recover from a failed single-step by requesting halt");
```

Reply:
> What have you please meant by "throw-away" halt? 

By "throw-away" halt I mean a halt request without any error processing. Here I wanted to emphasize that we have no guarantees that the halt actually happens.

Reply:
anyway. I've adopted you suggestion and just added an additional printout that checks for halt result. Could you please take another look?